src/com/palmergames/bukkit/towny/listeners/TownyPlayerListener.java

```diff
@@ -771,23 +776,80 @@ public void onPlayerChangePlotEvent(PlayerChangePlotEvent event) {
-		
-		Player player = event.getPlayer();		
-		WorldCoord to = event.getTo();
-		Resident resident = TownyUniverse.getInstance().getDataSource().getResident(player.getName());
 
-		if (to.getTownBlock().getTown().hasOutlaw(resident))
-			TownyMessaging.sendMsg(player, Translation.of("msg_you_are_an_outlaw_in_this_town", to.getTownBlock().getTown()));
+		Resident outlaw = TownyUniverse.getInstance().getDataSource().getResident(event.getPlayer().getName());
+		Town town = event.getEnteredtown();
+		TownBlock location = event.getTo().getTownBlock();
+		
+		String[] split = new String[0];
+		long cooldown = TownySettings.getOutlawKickCooldown();
+		
+		// If this resident is an outlaw in this town
+		if (location.getTown().hasOutlaw(outlaw)) {
+			if (!TownySettings.canOutlawsEnterTowns()) {
+				// Are towns warned when there is an outlaw entering?
+				if (TownySettings.doTownsGetWarnedOnOutlaw()) {
+					TownyMessaging.sendPrefixedTownMessage(town, Translation.of("msg_outlaw_town_notify").replace("%s", outlaw.getName()));
+				}
+				// Cooldown messages
+				if (cooldown < 1) {
+					TownyMessaging.sendMsg(outlaw, Translation.of("msg_outlaw_kick", town));
+				} else {
+					TownyMessaging.sendMsg(outlaw, Translation.of("msg_outlaw_kick_cooldown").replace("%c", TimeMgmt.formatCountdownTime(cooldown)).replace("%s", town.getFormattedName()));
+				}
+				
+				new BukkitRunnable() {
+					@Override
+					public void run() {
+						
+						if (town.hasOutlaw(outlaw)) {
+							
+							Player player = outlaw.getPlayer();
+							
+							// Outlaw has/doesn't have a town.
+							if (!outlaw.hasTown()) {
+								if (player.getBedSpawnLocation() == null) { 
+									player.teleport(town.getWorld().getSpawnLocation());
+								} else {
+									player.teleport(player.getBedLocation());
+								}
+							} else if (outlaw.hasTown()) {
+								try {
+									if (outlaw.getTown().hasSpawn()) {
+										try {
+											SpawnUtil.sendToTownySpawn(player, split, outlaw.getTown(), Translation.of("msg_err_cant_afford_tp"), false, false, SpawnType.TOWN);
+										} catch(TownyException exception) {
+											player.teleport(town.getWorld().getSpawnLocation());
+										}
+									} else {
+										try {
+											SpawnUtil.sendToTownySpawn(player, split, outlaw, Translation.of("msg_err_cant_afford_tp"), false, false, SpawnType.RESIDENT);
+										} catch(TownyException exception) {
+											player.teleport(town.getWorld().getSpawnLocation());
+										}
+									}
+								} catch (NotRegisteredException e) {
+									player.teleport(town.getWorld().getSpawnLocation());
+								}
+								TownyMessaging.sendMsg(outlaw, Translation.of("msg_outlaw_kicked", town));
+							}
+						}
+					}
+				}.runTaskLater(plugin, cooldown * 20);
+			}
+			else {
+				TownyMessaging.sendMsg(outlaw, Translation.of("msg_you_are_an_outlaw_in_this_town", town));
+			}
```

Comment:
I would probably put this at the top after line 794, ie:

```java
		/*
		 * Town does consider the player an outlaw.
		 */
		if (event.getTo().getTownBlock().getTown().hasOutlaw(outlaw)) {
			
			// Outlaws are allowed to roam in all towns freely.
			if (TownySettings.canOutlawsEnterTowns()) {
				TownyMessaging.sendMsg(outlaw, Translation.of("msg_you_are_an_outlaw_in_this_town", town));
			
			// Outlaws are teleported away if they spend too long in a town who has outlawed them.
			} else {
				// All of the new code.
			}
		}

```

Reply:
Added this way of handling this on the latest commit.